Composite consistency checker

fickludd opened the pull request:
Adds checks of composite indexes and uniqueness constraints to the consistency checker. This in essence means duplicating existing tests but applied on composite alternative schema objects.

Looks huge, but this is because it sits on #8984. Only commits https://github.com/neo4j/neo4j/pull/8993/commits/127392591c21f0b8558d63ed5ec10cb2f9c67194, https://github.com/neo4j/neo4j/pull/8993/commits/50512cffd7fa618fd8cacb96eda9404ac48d0d6b and https://github.com/neo4j/neo4j/pull/8993/commits/50a03bbd6f5f847a45e0a1c17ec89c2f1deb9168 are relevant.

fickludd committed:
Additional consistency checker test for composite indexes

fickludd committed:
Consistency checking of composite indexes

fickludd committed:
Test for duplicate composite uniqueness constraint rules + clean-up

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/report/ConsistencyReport.java:
```diff
@@ -184,13 +184,13 @@ void forCounts( CountsEntry countsEntry,
-        void indexedMultipleTimes( IndexRule index, Object propertyValue, long count );
+        void indexedMultipleTimes( IndexRule index, Object[] propertyValues, long count );
 
         @Documented( "There is another node in the unique index with the same property value." )
```
this message is a bit wrong now

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -233,4 +195,69 @@ private void checkProperty( NodeRecord record,
+            PropertyBlock propertyBlock = propertyWithKey( properties, indexPropertyIds[i] );
+            if ( propertyBlock == null )
+            {
+                throw new IllegalStateException( "We should have checked that the index and node match before coming " +
```
Exception message should be more descriptive in case if enduser will get it and should also include set of known property ids and requested id to give some possible insight what went wrong

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -233,4 +195,69 @@ private void checkProperty( NodeRecord record,
+        {
+            indexedNodeIds = reader.query( query );
+        }
+        catch ( IndexNotApplicableKernelException e )
```
What exactly IndexNotApplicableKernelException mean here? shouldn't we report that something went wrong instead of assuming that we do not have matches?

fickludd replied:
Ah, you're right! If we get that exception someone messed up the index implementation, and the only thing to do is likely to crash.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -70,139 +73,98 @@ public void check( NodeRecord record,
+                if ( nodeHasSchemaProperties( nodePropertyIds, indexPropertyIds, NO_SUCH_PROPERTY_KEY ) )
                 {
-                    verifyNodeCorrectlyIndexed( nodeId, propertyValue, engine, indexRule, reader );
+                    Object[] values = getPropertyValues( properties, indexPropertyIds );
```
if we will make properties map instead (id -> block) we can avoid here searching over property block again by ids that we have already extracted and check before

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/schema/NodeSchemaMatcher.java:
```diff
@@ -0,0 +1,109 @@
+        return true;
+    }
+
+    public interface SchemaMatchAction<SUPPLIER extends SchemaDescriptor.Supplier, EXCEPTION extends Exception>
```
Why this can't be a `Consumer` interface instead of custom `Action` thingy?

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -70,139 +73,98 @@ public void check( NodeRecord record,
-                else
+
+                int[] indexPropertyIds = indexRule.schema().getPropertyIds();
+                if ( nodeHasSchemaProperties( nodePropertyIds, indexPropertyIds, NO_SUCH_PROPERTY_KEY ) )
```
why do we need to pass NO_SUCH_PROPERTY_KEY here?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LookupFilter.java:
```diff
@@ -45,13 +45,40 @@
-            return PrimitiveLongCollections.filter( indexedNodeIds,
-                    new LookupBasedExactMatchPredicate( accessor, propertyKeyId,
-                            value ) );
+            LongPredicate combinedPredicate = nodeId ->
```
`Predicates.all`?

MishaDemianenko replied:
the same in other place in this class

fickludd replied:
I tried to solve this with `Predicates.all` as you suggest, but I couldn't make it better. This is because we need to fetch the property value for every predicate which means that we would need to map them and create lambdas or something.

MishaDemianenko reviewed (changes requested):
Couple of comments, also would be interesting to know what performance impact does this change brings

fickludd committed:
List->Map optimization

fickludd committed:
Throw on index not supporting exact composite query

fickludd committed:
Use standard ThrowingConsumer instead of custom interface

fickludd committed:
Added motivations for ignoring exceptions in exactIndexMatches versions

fickludd committed:
Let test allocate dynamic records

fickludd committed:
Fixed Object/Property bug in CC LookupFilter

MishaDemianenko commented:
@fickludd what is the status of this PR?

fickludd commented:
@MishaDemianenko Oh, you should have another go at it, I tried to address all comments. It keeps failing on some docs-thing that I don't think I caused.

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/report/ConsistencyReport.java:
```diff
@@ -184,13 +184,13 @@ void forCounts( CountsEntry countsEntry,
 
-        @Documented( "There is another node in the unique index with the same property value." )
-        void uniqueIndexNotUnique( IndexRule index, Object propertyValue, long duplicateNodeId );
+        @Documented( "There is another node in the unique index with the same property values." )
```
what if we will say `value(s)` instead since it will work for single property index as well?

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -235,4 +195,67 @@ private void checkProperty( NodeRecord record,
+        {
+            throw new RuntimeException( format(
+                    "Cannot continue %s, as index provider does not support exact composite query: %s",
+                    this.getClass().getSimpleName(), Arrays.toString( query ) ), e );
```
why do we need to print this.getClass().getSimpleName() ? it's not like we can't continue only this check right?
would it be more reasonable to continue check when we are in situation like that and report the failure?

MishaDemianenko commented on community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:
```diff
@@ -1057,12 +1092,30 @@ protected void transactionData( GraphStoreFixture.TransactionDataBuilder tx,
         return result;
     }
 
+    private static DynamicRecordAllocator schemaAllocator = new DynamicRecordAllocator()
```
please do not declare fields in a first suitable place Move it to the top to other fields

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/LookupFilter.java:
```diff
@@ -45,13 +47,43 @@
+                    {
+                        int propertyKeyId = predicate.propertyKeyId();
+                        Property property = accessor.getProperty( nodeId, propertyKeyId );
+                        if ( property.isDefined() && !predicate.test( ((DefinedProperty)property).value() ) )
```
why do we need this check for defined property?

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/checking/full/PropertyAndNodeIndexedCheck.java:
```diff
@@ -235,4 +195,67 @@ private void checkProperty( NodeRecord record,
+        return indexedNodeIds;
+    }
+
+    public static boolean nodeHasSchemaProperties(
```
this method can be private now right?

MishaDemianenko reviewed (changes requested):
Have couple of comments, also would be  nice to know how introduced changes affect performance of CC, since that is already quite critical part and we should not do that worse then it is now.
Another point would be super nice if you will include in PR description all introduced checks for composite indexes(basically what we are checking for them)

fickludd committed:
Fixed PR comments round two

fickludd commented:
@MishaDemianenko addressed comments, it's yours again. Also manually verified that the time taken to run `time neo4j-admin check-consistency ...` on a 400 Mb LDBC database is the same on this branch as a recent `3.2.0-alpha07` build.

MishaDemianenko commented:
@fickludd looks good, only need rebase , will merge it on green
